protocol/common/message_sender.go

```diff
@@ -1224,15 +1226,35 @@ func (s *MessageSender) JoinPublic(id string) (*transport.Filter, error) {
 	return s.transport.JoinPublic(id)
 }
 
-// AddEphemeralKey adds an ephemeral key that we will be listening to
-// note that we never removed them from now, as waku/whisper does not
-// recalculate topics on removal, so effectively there's no benefit.
-// On restart they will be gone.
-func (s *MessageSender) AddEphemeralKey(privateKey *ecdsa.PrivateKey) (*transport.Filter, error) {
+func (s *MessageSender) getRandomEphemeralKey() *ecdsa.PrivateKey {
+	k := rand.Intn(len(s.ephemeralKeys)) //nolint: gosec
+	for _, key := range s.ephemeralKeys {
+		if k == 0 {
+			return key
+		}
+		k--
+	}
+	return nil
+}
+
+func (s *MessageSender) GetEphemeralKey() (*ecdsa.PrivateKey, error) {
 	s.ephemeralKeysMutex.Lock()
+	if len(s.ephemeralKeys) >= maxMessageSenderEphemeralKeys {
+		return s.getRandomEphemeralKey(), nil
```

Comment:
We could just use map's iterator to get a random since golang's map iterator is supposed to follow random order.

Reply:
I think the order is not specified officially, not random (I believe the implementation currently use random order, but that in theory might change), so this in theory is a bit safer in case things change